### PR TITLE
adds control server acceleartion on menu open event, adds more failsafes for shutting down desktop when parent gone

### PR DESCRIPTION
**Pull request by James-Pickett**
This PR repurposes the "monitor_server" that was used only for the desktop user process to check on the root processes' health. The new runner server has an additional endpoint for "menuopened" that the desktop user process can call when it gets a notification from systray that the menu has been opened.

It introduces RunnerServer and renames DesktopServer to UserServer. Both servers are related to the desktop but one is run by the DesktopRunner and one is run by the Desktop User Process.

After this PR merges, I will do some reshuffling of the desktop package structure and put packages into either a `user` or `runner` folder based on their role.

It also:

* Updates DesktopRunner to DeRegister desktop user clients form the RunnerServer when they are assumed to be dead. This way if the assumption was false and a new desktop user process was spun up, the old one should terminate it self since the RunnerServer will stop responding to it.
* Updates the Desktop User Process to self terminate if it's ppid is less than 2. We've seen some funny cases where a user has 2 desktop processes running and one has a parent of 1 or 0.

closes: https://github.com/kolide/launcher/issues/1155
closes: https://github.com/kolide/launcher/issues/1141

### CODE REVIEW

File: cmd/launcher/desktop.go [hunk 1]
```diff
@@ -18,8 +17,11 @@ import (
 	"github.com/kolide/kit/ulid"
 	"github.com/kolide/launcher/ee/desktop/menu"
 	"github.com/kolide/launcher/ee/desktop/notify"
-	"github.com/kolide/launcher/ee/desktop/server"
+	rootserver "github.com/kolide/launcher/ee/desktop/runner/server"
+	desktopserver "github.com/kolide/launcher/ee/desktop/server"
```
Comment (James-Pickett) [inline cmd/launcher/desktop.go:21]: wonder if we should just go ahead an name these packages `rootserver` and `desktopserver`

File: ee/desktop/runner/server/server.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
+package server
```
Comment (James-Pickett) [inline ee/desktop/runner/server/server.go:1]: should we call this `runnerserver` or `rootserver`?

File: ee/desktop/runner/server/server.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+		}
+	})
+
+	// menu opened endpoint
+	mux.HandleFunc(MenuOpenedEndpoint, func(w http.ResponseWriter, r *http.Request) {
+		if r.Body != nil {
+			r.Body.Close()
+		}
+
+		ms.knapsack.SetControlRequestIntervalOverride(5*time.Second, 1*time.Minute)
```
Comment (James-Pickett) [inline ee/desktop/runner/server/server.go:64]: hooray for `Knaspsack`, it made this part so easy

File: ee/desktop/runner/server/server.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	"github.com/kolide/launcher/pkg/agent/types"
+)
+
+type RootServer struct {
+	server                *http.Server
+	listener              net.Listener
+	logger                log.Logger
+	desktopProcAuthTokens map[string]string
+	mutex                 sync.Mutex
+	knapsack              types.Knapsack
```
Comment (directionless) [inline ee/desktop/runner/server/server.go:24]: Optional, but you can declare what you need in an interface. And not the whole knapsack

File: cmd/launcher/desktop.go (hunk 1 above)
Comment (James-Pickett) [reply cmd/launcher/desktop.go:21]: rootserver is the server run by the root process

desktopserver is the server run by the desktop process

maybe we should call them `rootprocserver` and `desktopprocserver`

any suggestions?

File: cmd/launcher/desktop.go [hunk 5]
```diff
@@ -75,6 +82,10 @@ func runDesktop(args []string) error {
 		"pid", os.Getpid(),
 	)
 
+	if *fldebug {
+		level.AllowDebug()
+	}
```
[the remark this answers is not in the document]
Comment (James-Pickett) [reply cmd/launcher/desktop.go:87]: you correct and this is actually set earlier when we create the logger, removing

File: cmd/launcher/desktop.go [hunk 6]
```diff
@@ -32,20 +34,25 @@ func runDesktop(args []string) error {
 			"",
 			"hostname launcher is connected to",
 		)
-		flauthtoken = flagset.String(
-			"authtoken",
+		flUserServerAuthToken = flagset.String(
+			"user_server_auth_token",
 			"",
-			"auth token for desktop server",
+			"auth token for user user server",
```
Comment (RebeccaMahany) [inline cmd/launcher/desktop.go:40]: I think typo 🙂 

```suggestion
			"auth token for user server",
```

File: ee/desktop/runner/server/server.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kolide/kit/ulid"
+)
+
+type RunnerServer struct {
```
Comment (RebeccaMahany) [inline ee/desktop/runner/server/server.go:17]: Mind adding a short comment about what the runner server's responsibilities are? `Allows user desktop processes to monitor + notify parent process` or similar

File: ee/desktop/runner/server/server.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+		}
+	})
+
+	// menu opened endpoint
+	mux.HandleFunc(MenuOpenedEndpoint, func(w http.ResponseWriter, r *http.Request) {
+		if r.Body != nil {
+			r.Body.Close()
+		}
+
+		controlRequestIntervalOverrider.SetControlRequestIntervalOverride(5*time.Second, 1*time.Minute)
```
Comment (RebeccaMahany) [inline ee/desktop/runner/server/server.go:67]: Could we make these durations consts so it's easier to identify/update them in the future?

**Review by RebeccaMahany — APPROVED** [review]
Just a couple small comments, LGTM overall!

File: cmd/launcher/desktop.go (hunk 6 above)
Comment (James-Pickett) [reply cmd/launcher/desktop.go]: done

File: ee/desktop/runner/server/server.go (hunk 7 above)
Comment (James-Pickett) [reply ee/desktop/runner/server/server.go:19]: done

File: ee/desktop/runner/server/server.go (hunk 8 above)
Comment (James-Pickett) [reply ee/desktop/runner/server/server.go]: done

**Review by RebeccaMahany — APPROVED** [review]
🔥

File: ee/desktop/runner/server/server.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+const (
+	HealthCheckEndpoint                = "/health"
+	MenuOpenedEndpoint                 = "/menuopened"
+	controlRequestAccelerationInterval = 5 * time.Second
+	controlRequestAcclerationDuration  = 1 * time.Minute
+)
+
+type controlRequestIntervalOverrider interface {
```
Comment (seejdev) [inline ee/desktop/runner/server/server.go:35]: Yay, cool to see the knapsack sub-interface in action. This feels clean too

File: ee/desktop/runner/server/server.go [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+	SetControlRequestIntervalOverride(time.Duration, time.Duration)
+}
+
+func New(logger log.Logger, controlRequestIntervalOverrider controlRequestIntervalOverrider) (*RunnerServer, error) {
+	listener, err := net.Listen("tcp", "localhost:0")
```
Comment (seejdev) [inline ee/desktop/runner/server/server.go:40]: Does using port zero mean it will choose a dynamically generated unused port?

**Review by seejdev — APPROVED** [review]

File: ee/desktop/runner/server/server.go (hunk 10 above)
Comment (James-Pickett) [reply ee/desktop/runner/server/server.go:40]: yea, it just picks a random port, it's what the current "monitor server" that is doing now on stable

**Pull request merged**